File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
'''
1. create a hashMap to count frequency of elements in nums
res = []
2. bucket = create a new array and initialize it with len(nums) + 1 empty arrays
3. for key, value in hashMap:
        if len(bucket[value]) > 0:
            bucket[value].push(key)
        else:
            bucket[value] = [key]
4. Loop in reverse order over bucket:
        put elements from bucket while k > 0
'''

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashMap = {}
        res = []

        bucket = []

        for i in range (len(nums) + 1):
            bucket.append([])
        
        for n in nums:
            hashMap[n] = hashMap.get(n, 0) + 1
        
        for key, val in hashMap.items():
                bucket[val].append(key)

        for i in range(len(bucket) - 1, -1, -1):
            if(len(bucket[i]) > 0):
                for n in bucket[i]:
                    res.append(n)
                    k-=1
                    if k == 0:
                        return res

        return res
                
                

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */

 /*
 Use bucket sort
 1. hashMap = {count frequency of each num in nums}
    res = []
 2. bucket = [].size(nums.length + 1).fill([])
 3. for key, value of hashMap.entries() {
        bucket[value].push(key)
    }
 4. find top k elements
    for (var i = bucket.length - 1; i>=0; i--) {
        if (bucket[i].length > 0) {
            for (var n in bucket[i]) {
                res.push(n)
                if (res.length == k) {
                    return res
                }
            }
        }
    }
 */
var topKFrequent = function(nums, k) {
    var hashMap = {}
    var res = []

    // var bucket = new Array(nums.length + 1).fill([])
    var bucket = []

    for (var i = 0; i <=nums.length; i++) {
        bucket[i] = []
    }

    for (var n of nums) {
        hashMap[n] = (hashMap[n] || 0) + 1
    }

    for (var [key, value] of Object.entries(hashMap)) {
        bucket[value].push(key)
    }

    for (var i = bucket.length - 1; i >=0; i--) {
        if (bucket[i].length > 0) {
            for (var n of bucket[i]) {
                res.push(parseInt(n))
            }
            if (res.length == k) {
                return res
            }
        }
    }
};